cms/users/models.py

```diff
@@ -1,13 +1,90 @@
-from django.contrib.auth.models import AbstractUser
+from collections.abc import Iterable
+
+from django.conf import settings
+from django.contrib.auth.models import AbstractUser, Group
 from django.db import models
 from django.utils.functional import cached_property
 
+from cms.teams.models import Team
+
 
 class User(AbstractUser):
     """Barebones custom user model."""
 
+    # Unique identifier for user lookup, as emails (used as usernames) may change.
+    external_user_id = models.UUIDField(unique=True, null=True, editable=False)
     teams = models.ManyToManyField("teams.Team", related_name="users", blank=True)
 
+    @property
+    def is_external_user(self) -> bool:
+        """Return True for users who originate from an external identity provider
+        (i.e. they have an external_user_id) and have no usable password set.
+        External accounts rely on third-party authentication, so they should
+        never possess a native password.
+        """
+        return self.external_user_id is not None and not self.has_usable_password()
+
+    def assign_groups_and_teams(self, /, groups_ids: Iterable[str]) -> None:
+        """Assign groups and teams to the user based on their Cognito groups."""
+        self._assign_groups(groups_ids)
+        self._assign_teams(groups_ids)
+
+        self.save()
```

Comment:
Nit: Setting groups and teams doesn't change anything on the model, so this change is unnecessary

Reply:
This step is necessary to save the changes made by the `update_details` call before this function, which edits the model itself. If it makes the code easier to follow, we can move the `self.save()` method out of this function and place it in the middleware after `assign_groups_and_teams`.

Given it was removed and no test failed around user's details change, it implies we have a gap in the test. (Let's add in the follow up).

Reply:
Thanks @MebinAbraham. I agree with moving the `self.save()` method out of this function and place it after `assign_groups_and_teams ` in the middleware where it makes more sense.

I thought I had covered this test case, =/. Let me have a quick look if not will add to it to follow-up.